File: deepspeed/inference/v2/kernels/ragged_ops/top_k_gating/top_k_gating.py
# Copyright (c) Microsoft Corporation.
# SPDX-License-Identifier: Apache-2.0

# DeepSpeed Team

import torch

from typing import Tuple

from ... import DSKernelBase
from ....inference_utils import DtypeEnum
from ....ragged import RaggedBatchWrapper
from deepspeed.ops.op_builder import RaggedOpsBuilder


class RaggedTopKGating(DSKernelBase):
    """
    CUDA implementation of top-1 gating. This will perform a softmax on the logits,
    and return the scale as well as its idx within that expert's allocation.
    """

    supported_logit_dtypes = [DtypeEnum.fp16, DtypeEnum.bf16, DtypeEnum.fp32]

    def __init__(self, logit_dtype: DtypeEnum) -> None:

        if not isinstance(logit_dtype, DtypeEnum):
            logit_dtype = DtypeEnum(logit_dtype)

        if logit_dtype not in RaggedTopKGating.supported_logit_dtypes:
            raise RuntimeError(f"Unsupported logit dtype {logit_dtype}")

        inf_module = RaggedOpsBuilder().load()
        self.kernel = inf_module.top_k_gating

    def __call__(self, expert_counts: torch.Tensor, scores: torch.Tensor, assignments: torch.Tensor,
                 offsets: torch.Tensor, logits: torch.Tensor,
                 batch: RaggedBatchWrapper) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor, torch.Tensor]:
        """
        Perform the ragged top_k_gating.

        Arguments:
            expert_counts (torch.Tensor): Tensor of 0s of shape [n_experts] to be filled with
                number of tokens assigned to each expert. This must be filled with 0s else
                the copy kernel will buffer overflow. In order to minimize the zero-fill cost,
                it is recommended to write to 0 during the MoE output remapping.
            scores (torch.Tensor): Preallocated output of shape [n_tokens, n_top_k] to place expert scaling
                value.
            expert_assignment (torch.Tensor): Preallocated output of shape [n_tokens, n_top_k] to place
                which expert a token has been assigned to.
            expert_offset (torch.Tensor): Preallocated output of shape [n_tokens, n_top_k] to place which
                offset within an experts group a token is.
            logits (torch.Tensor): Raw logits of gating function.
            batch (RaggedBatchWrapper): Batch information for ragged tensor.

        Returns:
            tuple of (expert_counts, scores, expert_assignment, expert_offset)
        """
        self.kernel(expert_counts, scores, assignments, offsets, logits, batch.batch_metadata_buffer())
        return expert_counts, scores, assignments, offsets

File: deepspeed/inference/v2/kernels/ragged_ops/top_k_gating/top_k_gating.cpp
// Copyright (c) Microsoft Corporation.
// SPDX-License-Identifier: Apache-2.0

// DeepSpeed Team

#include "top_k_gating.h"
#include <c10/cuda/CUDAStream.h>

#define DISPATCH_TOP_K_GATING(T_TYPE, C_TYPE)                   \
    if (logits.options().dtype() == torch::T_TYPE) {            \
        launch_top_k_gating((int32_t*)expert_counts.data_ptr(), \
                            (float*)scores.data_ptr(),          \
                            (int32_t*)assignments.data_ptr(),   \
                            (int32_t*)offsets.data_ptr(),       \
                            (const C_TYPE*)logits.data_ptr(),   \
                            batch_metadata_ptr,                 \
                            n_tokens,                           \
                            n_experts,                          \
                            n_top_k,                            \
                            at::cuda::getCurrentCUDAStream());  \
        return;                                                 \
    }

/*
Perform softmax plus atomics in order to do first pass of top_k_gating.
*/
void top_k_gating(torch::Tensor& expert_counts,
                  torch::Tensor& scores,
                  torch::Tensor& assignments,
                  torch::Tensor& offsets,
                  torch::Tensor& logits,
                  torch::Tensor& batch_metadata)
{
    const int32_t n_tokens = scores.size(0);
    const int32_t n_top_k = scores.size(1);

    // Should have the same buffer size for scores, offsets, and assignments
    TORCH_CHECK(n_tokens == offsets.size(0));
    TORCH_CHECK(n_tokens == logits.size(0));
    TORCH_CHECK(n_tokens == assignments.size(0));

    TORCH_CHECK(n_top_k == offsets.size(1));
    TORCH_CHECK(n_top_k == assignments.size(1));

    TORCH_CHECK(expert_counts.scalar_type() == torch::kInt32);
    TORCH_CHECK(scores.scalar_type() == torch::kFloat);
    TORCH_CHECK(assignments.scalar_type() == torch::kInt32);
    TORCH_CHECK(offsets.scalar_type() == torch::kInt32);

    const int32_t n_experts = logits.size(1);
    const RaggedBatchDescriptor* batch_metadata_ptr =
        reinterpret_cast<const RaggedBatchDescriptor*>(batch_metadata.data_ptr());

    DISPATCH_TOP_K_GATING(kFloat, float)
    DISPATCH_TOP_K_GATING(kHalf, __half)
#ifdef BF16_AVAILABLE
    DISPATCH_TOP_K_GATING(kBFloat16, __nv_bfloat16)
#endif

    TORCH_CHECK(false, "Unsupported dtype for logits in top_k_gating");
}
